service/history/conflictResolverV2.go

```diff
@@ -0,0 +1,191 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package history
+
+import (
+	"github.com/uber/cadence/.gen/go/shared"
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/cluster"
+	"github.com/uber/cadence/common/collection"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/log/tag"
+	"github.com/uber/cadence/common/persistence"
+)
+
+const (
+	defaultPageSize = 100
+)
+
+type (
+	conflictResolverV2 interface {
+		reset(prevRunID string, requestID string, replayEventID int64, info *persistence.WorkflowExecutionInfo) (mutableState, error)
+	}
+
+	conflictResolverV2Impl struct {
+		shard           ShardContext
+		clusterMetadata cluster.Metadata
+		context         workflowExecutionContext
+		historyMgr      persistence.HistoryManager
+		historyV2Mgr    persistence.HistoryV2Manager
+		logger          log.Logger
+	}
+)
+
+func newConflictResolverV2(shard ShardContext, context workflowExecutionContext, historyMgr persistence.HistoryManager, historyV2Mgr persistence.HistoryV2Manager,
+	logger log.Logger) *conflictResolverImpl {
+
+	return &conflictResolverImpl{
+		shard:           shard,
+		clusterMetadata: shard.GetService().GetClusterMetadata(),
+		context:         context,
+		historyMgr:      historyMgr,
+		historyV2Mgr:    historyV2Mgr,
+		logger:          logger,
+	}
+}
+
+func (r *conflictResolverV2Impl) reset(prevRunID string, requestID string, replayEventID int64, info *persistence.WorkflowExecutionInfo, localVersionHistories persistence.VersionHistories) (mutableState, error) {
+	domainID := r.context.getDomainID()
+	execution := *r.context.getExecution()
+	startTime := info.StartTimestamp
+	eventStoreVersion := info.EventStoreVersion
+	branchToken := info.GetCurrentBranch()
+	replayNextEventID := replayEventID + 1
+
+	var resetMutableStateBuilder *mutableStateBuilder
+	var sBuilder stateBuilder
+	var lastEvent *shared.HistoryEvent
+	var lastFirstEventID int64
+
+	eventsToApply := replayNextEventID - common.FirstEventID
+
+	iter := collection.NewPagingIterator(func(paginationToken []byte) ([]interface{}, []byte, error) {
+		var paginateItems []interface{}
+		historyBatch, _, token, err := r.getHistoryBatch(domainID, execution, common.FirstEventID, replayNextEventID, paginationToken, eventStoreVersion, branchToken)
+		for _, history := range historyBatch {
+			paginateItems = append(paginateItems, history)
+		}
+		return paginateItems, token, err
+	})
+
+	for iter.HasNext() {
+		item, err := iter.Next()
+		if err != nil {
+			r.logger.Error("Conflict resolution err getting history.", tag.Error(err))
+			return nil, err
+		}
+
+		history := item.(*shared.History).Events
+		batchSize := int64(len(history))
```

Comment:
the verification logic here is not correct, should assert on last event on the last batch has the event ID == replayEventID